bpf/lib/proxy.h

```diff
@@ -235,17 +235,107 @@ ctx_redirect_to_proxy6(struct __ctx_buff *ctx, void *tuple __maybe_unused,
 }
 #endif /* ENABLE_IPV6 */
 
+#ifdef ENABLE_TPROXY
+/**
+ * extract_tuple4 / extract_tuple6
+ *
+ * Extracts the packet 5-tuple into 'tuple'.
+ *
+ * Note that it doesn't fully initialize 'tuple' as the directionality bit
+ * is unused in the proxy paths.
+ */
+#define IP_TUPLE_EXTRACT_FN(NAME, PREFIX)				\
+static __always_inline int						\
+NAME(struct __ctx_buff *ctx, struct PREFIX ## _ct_tuple *tuple)		\
+{									\
+	int err, l4_off;						\
+									\
+	err = PREFIX ## _extract_tuple(ctx, tuple, &l4_off);		\
+	if (err != CTX_ACT_OK)						\
+		return err;						\
+									\
+	/**								\
+	 * TODO: Consider passing ports with meta/mark from		\
+	 * __ctx_redirect_to_proxy() to avoid touching packet here	\
+	 */								\
+	if (ctx_load_bytes(ctx, l4_off, &tuple->dport, 4) < 0)		\
```

Comment:
Yeah, L4 is a little bit trickier just because you have to figure out the l4 offset so I figured maybe we can optimize that part out at least. But I don't know that it's actually worth it. I have no immediate intention to address this TODO, so perhaps I should just drop it.